src/core/connection.c

```diff
@@ -1892,6 +1892,225 @@ QuicConnRestart(
+    QuicCopyMemory(TicketCursor, EncodedLocalTP, EncodedTransportParametersLength);
+    TicketCursor += EncodedTransportParametersLength;
+    TicketCursor = QuicVarIntEncode(AppDataLength, TicketCursor);
+    if (AppDataLength > 0) {
+        QuicCopyMemory(TicketCursor, AppResumptionData, AppDataLength);
+    }
+
+    Status =
+        QuicTlsSendTicket(Connection->Crypto.TLS, TotalTicketLength, TicketBuffer);
+
+Error:
+    if (TicketBuffer != NULL) {
+        //
+        // TODO: This might need to be kept around for longer depending on TLS implementation...
+        //
+        QUIC_FREE(TicketBuffer);
+    }
+
+    if (EncodedLocalTP != NULL) {
+        QUIC_FREE(EncodedLocalTP);
+    }
+
+    if (AppResumptionData != NULL) {
+        QUIC_FREE(AppResumptionData);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+BOOLEAN
+QuicConnRecvResumptionTicket(
+    _In_ QUIC_CONNECTION* Connection,
+    _In_ uint16_t TicketLength,
+    _In_reads_(TicketLength)
+        const uint8_t* Ticket
+    )
+{
+    BOOLEAN ResumptionAccepted = FALSE;
+    if (QuicConnIsServer(Connection)) {
+        uint16_t Offset = 0;
+        QUIC_VAR_INT TicketVersion = 0, AlpnLength = 0, TPLength = 0, AppTicketLength = 0;
+        if (!QuicVarIntDecode(TicketLength, Ticket, &Offset, &TicketVersion)) {
+            QuicTraceEvent(
+                ConnError,
+                "[conn][%p] ERROR, %s.",
+                Connection,
+                "Resumption Ticket version failed to decode");
+            goto Error;
+        }
+        if (TicketVersion != QUIC_TLS_RESUMPTION_TICKET_VERSION) {
+            QuicTraceEvent(
+                ConnError,
+                "[conn][%p] ERROR, %s.",
+                Connection,
+                "Resumption Ticket version unsupported");
+            goto Error;
+        }
+
+        uint32_t QuicVersionHost = QuicByteSwapUint32(*(uint32_t*)(Ticket + Offset));
+        if (!QuicIsVersionSupported(QuicVersionHost)) {
+            QuicTraceEvent(
+                ConnError,
+                "[conn][%p] ERROR, %s.",
+                Connection,
+                "Resumption Ticket for unsupported QUIC version");
+            goto Error;
+        }
+        Offset += sizeof(QuicVersionHost);
+
+        if (!QuicVarIntDecode(TicketLength, Ticket, &Offset, &AlpnLength)) {
+            QuicTraceEvent(
+                ConnError,
+                "[conn][%p] ERROR, %s.",
+                Connection,
+                "Resumption Ticket ALPN length failed to decode");
+            goto Error;
+        }
+        if (QuicTlsAlpnFindInList(
+                Connection->Session->AlpnListLength, Connection->Session->AlpnList,
+                (uint8_t)AlpnLength, Ticket + Offset) == NULL) {
+            QuicTraceEvent(
+                ConnError,
+                "[conn][%p] ERROR, %s.",
+                Connection,
+                "Resumption Ticket ALPN not present in ALPN list");
+            goto Error;
+        }
```

Comment:
Would it ever make sense to allow for alpn upgrade? As in, you have a ticket from abc-00 and you use it, but connect with abc-01? I'm not sure it would make sense...

Reply:
I don't think that would make sense. The application would ultimately have to make the call, not us, if allowed. However, my goal here is to make sure the token presented actually matches the server.
Suppose a deployment is doing some A/B testing and has a load balancer distribute traffic between servers running `abc-00` and `abc-01`, unless the `abc` protocol is carefully designed, tickets for one version could confuse/break the other version.